src/metabase/upload.clj

```diff
@@ -533,59 +539,75 @@
+                                      remove-auto-pk-cols)
+        normalized-header (map first normalized-header+header)
         [extra missing _both] (data/diff (set normalized-header)
-                                         (set normalized-field-names))]
-
+                                         ;; ignore the auto-pk field, it will be filled automatically
+                                         (disj (set normalized-field-names) auto-pk-column-name))]
+    ;; check for duplicates
+    (when (some #(< 1 %) (vals (frequencies normalized-header)))
+      (throw (ex-info (tru "The CSV file contains duplicate column names.")
+                      {:status-code 422})))
     (if (and (empty? extra) (empty? missing))
       {:parse-rows (fn [rows]
-                     (let [rows' rows ;; TODO: remove auto-pk columns here
+                     (let [rows' (cond->> rows
+                                   (seq auto-pk-col-indices)
+                                   (map remove-auto-pk-cols))
                            col-upload-types (map (comp base-type->upload-type :base_type fields-by-normed-name)
                                                  normalized-header)]
                        (parse-rows* col-upload-types rows')))
-       :table-col-names (map (comp :name fields-by-normed-name) normalized-header)}
+       :table-col-names (map (comp :name fields-by-normed-name) normalized-header)
+       :table-has-auto-pk? (contains? fields-by-normed-name auto-pk-column-name)}
       (throw (ex-info (tru "The schema of the CSV file does not match the schema of the table.")
                       {:status-code 422})))))
 
 (defn- append-csv!*
   [database table file]
   (with-open [reader (bom/bom-reader file)]
     (let [[header & rows] (csv/read-csv reader)
-          {:keys [parse-rows table-col-names]} (check-schema table header)]
+          {:keys [parse-rows table-col-names table-has-auto-pk?]} (check-schema table header)
+          driver (driver.u/database->driver database)
+          auto-pk-type (driver/upload-type->database-type driver ::auto-incrementing-int-pk)
+          create-auto-pk? (not table-has-auto-pk?)]
       (try
         (let [parsed-rows (parse-rows rows)]
-          (driver/insert-into! (driver.u/database->driver database)
+          (when create-auto-pk?
+            (driver/add-columns! driver
+                                 (:id database)
+                                 (table-identifier table)
+                                 {:_mb_row_id auto-pk-type}))
+          (driver/insert-into! driver
                                (:id database)
                                (table-identifier table)
                                table-col-names
                                parsed-rows)
+          (scan-and-sync-table! database table)
+          (when create-auto-pk?
+            (let [auto-pk-field (get-auto-pk-column (:id table))]
```

Comment:
didn't we have a comment somewhere explaining why we need to do this? could you keep it here?

Reply:
also, wdyt if we handle this in `scan-and-sync-table!`?

Reply:
Leaving a comment is a good idea, but I don't think we should put this in `scan-and-sync-table!` because they seem logically separate, they happen in different contexts for different reasons. I actually want to move `scan-and-sync-table!` to a sync namespace, as we discussed on the last PR. And there is a difference between the two cases it's used: when we create the CSV table the auto-pk column is updated unconditionally, and here it's updated only on the condition that it wasn't present before.